File: code/dask/run_fr_dask.m
% code/dask/run_fr_dask.m
% {{{
% This code accompanies the book _Python for MATLAB Development:
% Extend MATLAB with 300,000+ Modules from the Python Package Index_ 
% ISBN 978-1-4842-7222-0 | ISBN 978-1-4842-7223-7 (eBook)
% DOI 10.1007/978-1-4842-7223-7
% https://github.com/Apress/python-for-matlab-development
% 
% Copyright © 2022 Albert Danial
% 
% MIT License:
% Permission is hereby granted, free of charge, to any person obtaining a copy
% of this software and associated documentation files (the "Software"), to deal
% in the Software without restriction, including without limitation the rights
% to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
% copies of the Software, and to permit persons to whom the Software is
% furnished to do so, subject to the following conditions:
% 
% The above copyright notice and this permission notice shall be included in
% all copies or substantial portions of the Software.
% 
% THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
% IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
% FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
% THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
% LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
% FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
% DEALINGS IN THE SOFTWARE.
% }}}
clear all
Im = @py.importlib.import_module;
np      = Im('numpy');
pysolve = Im('pysolve');

T_main_s = tic;

%%
load('KCMb_7.mat') % K, C, M, b, keep_dof
fprintf('Keeping response at %d DOF\n', length(keep_dof))
load_history = load('load_hist_smoothed.txt');
n_jobs = int64(100);
n_FFT = 4096;
solve_with = "coiled";
% solve_with = "localhost";
% solve_with = "digitalocean";

%%
n_freq = int64(floor(n_FFT/2));
dT    = 5.0e-4;   % time between each load_history point
Time  = 0:dT:dT*n_FFT;
Fs    = 1/dT;     % sampling frequency Hz
Hertz = linspace(0, Fs/2, n_freq); % only to Nyquist
dHz   = Hertz(2);
P = fft(load_history);
P_Nyq = P(1:n_freq);
n_dof = size(M,1);
fprintf('Time: %d times, 0.0 to %.6f sec steps of %.6f sec\n', n_FFT, Time(end), dT)
fprintf('Freq: %d freq, 0.0 to %.1f Hz steps of %.1f Hz\n', n_freq, Hertz(end), dHz)

% convert MATLAB-native arrays to Python-native
b = rand(n_dof,1);
K = mat2py(K);
C = K * 0.004; % proportional damping
M = mat2py(M);
b = np.array(b);
P_Nyq = mat2py(P_Nyq);
Hertz = np.array(Hertz);
keep_dof = np.array(keep_dof).astype(np.int64);
  
x = pysolve.remote_solve(solve_with, n_jobs, K, C, M, b, P_Nyq, Hertz, keep_dof);
x = py2mat(x);

fprintf('%d frequencies took %.3f s\n', n_freq, toc(T_main_s))

File: code/dask/run_fr_dask.py
#!/usr/bin/env python3
# code/dask/run_fr_dask.py
# {{{
# This code accompanies the book _Python for MATLAB Development:
# Extend MATLAB with 300,000+ Modules from the Python Package Index_ 
# ISBN 978-1-4842-7222-0 | ISBN 978-1-4842-7223-7 (eBook)
# DOI 10.1007/978-1-4842-7223-7
# https://github.com/Apress/python-for-matlab-development
# 
# Copyright © 2022 Albert Danial
# 
# MIT License:
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
# 
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
# 
# THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
# DEALINGS IN THE SOFTWARE.
# }}}
import sys
import numpy as np
from   timeit import default_timer
from run_fr import read_tri, stiffness_terms, save_to_pkl, \
    constrain_dof, toc, global_matrices
import dask
from dask.distributed import Client, LocalCluster
import pysolve

tic = default_timer

def main():
    T_total_s = tic()
    model = 'one'
    model = 'tiny'
    model = 'sat.0'
    model = 'satellite/satellite.7'
    node_xy = read_tri(f'{model}.ele', f'{model}.node')[1]
    solve_with = 'localhost' # localhost digitalocean coiled
    solve_with = 'coiled' # localhost digitalocean coiled
    n_jobs = 3
    n_jobs = 2

    load_history = np.loadtxt('data/load_hist_smoothed.txt')
    n_FFT  = 8192
    n_FFT  = 4096
    n_FFT  = 2048
    n_freq = n_FFT // 2 # + 1 # only work up to Nyquist frequency
    dT = 5.0e-4  # time between each load_history point
    Time  = np.arange(0, dT*n_FFT, dT)
    Fs = 1/dT    # sampling frequency
    Hertz = np.linspace(0, Fs/2, num=n_freq) # only to Nyquist
    dHz = Hertz[1]
    print(f'Time: {n_FFT} times, 0.0 to {Time[-1]:.6f} sec steps of {dT:.6f} sec')
    print(f'Freq: {n_freq} freq, 0.0 to {Hertz[-1]:.1f} Hz steps of {dHz:.1f} Hz')
    print(f'      {n_jobs} workers')

    K, C, M, b, u_to_c = global_matrices(model)
    P = np.fft.rfft(load_history, n=n_FFT)
    P_Nyq = P[:n_freq] # only to Nyquist

    keep_nodes = np.argwhere((np.abs(23.5 - node_xy[:,0]) < 1.5) *
                             (np.abs(10.5 - node_xy[:,1]) < 1.5)).ravel()
    keep_dof_full = sorted(np.hstack([keep_nodes*2, keep_nodes*2+1]))
    keep_dof = sorted([u_to_c[_] for _ in keep_dof_full]) # constrained DOF

    if not len(keep_dof):
        print('keep_dof is empty; expand radius')
        return
    print(f'keeping {len(keep_dof)} out of {K.shape[0]} DOF')

    solnx = pysolve.remote_solve(solve_with, n_jobs, K, C, M, b,
                                 P_Nyq, Hertz, keep_dof)
    print(solnx)
    toc('Total time = ', T_total_s)

if __name__ == "__main__": main()
